src/adjoint.jl

```diff
@@ -1,16 +1,39 @@
 using MacroTools
 using MacroTools: @q, combinedef
+using ChainRulesCore: AbstractZero, Zero, DoesNotExist
 
 named(arg) = isexpr(arg, :(::)) && length(arg.args) == 1 ? :($(gensym())::$(arg.args[1])) : arg
 
 typeless(x) = MacroTools.postwalk(x -> isexpr(x, :(::), :kw) ? x.args[1] : x, x)
 isvararg(x) = isexpr(x, :(::)) && namify(x.args[2]) == :Vararg
 
+"""
+    generic2diff(x)
+
+Convert input `x` from the internal Zygote format to the ChainRules differential types.
+"""
+generic2diff(x) = x
+generic2diff(::Nothing) = Zero()
+generic2diff(t::Union{Tuple, NamedTuple}) = map(generic2diff, t)
+
+"""
+    diff2generic(x)
+
+Convert input `x` from the ChainRules differential types to the internal Zygote format.
+"""
+diff2generic(x) = x
+diff2generic(::AbstractZero) = nothing
+diff2generic(t::Union{Tuple, NamedTuple}) = map(diff2generic, t)
+diff2generic(::Nothing) =
+    @warn "Use of 'nothing' to represent zero gradients is deprecated, " *
+    "use Zero() or DoesNotExist() from ChainRules";
+    nothing
```

Comment:
Is this one needed?
I don't think it is.
That would mean that internally Zygote was using a `nothing`.

Should we change this to be a warning like:
```suggestion
function diff2generic(::Nothing)
    @error "Zygote internal use  of 'nothing', rather than `AbstractZero`, detected.  This should never occur. Please open an issue on https://github.com/FluxML/Zygote.jl/issues, including the full text of this message"  stacktrace=stacktrace()
    return nothing
end
```